kotlinx-coroutines-core/common/src/channels/ConflatedChannel.kt

```diff
@@ -15,80 +13,118 @@ import kotlinx.coroutines.internal.*
  * Sender to this channel never suspends and [offer] always returns `true`.
  *
  * This channel is created by `Channel(Channel.CONFLATED)` factory function invocation.
- *
- * This implementation is fully lock-free.
  */
 internal open class ConflatedChannel<E> : AbstractChannel<E>() {
-    protected final override val isBufferAlwaysEmpty: Boolean get() = true
-    protected final override val isBufferEmpty: Boolean get() = true
-    protected final override val isBufferAlwaysFull: Boolean get() = false
+    protected final override val isBufferAlwaysEmpty: Boolean = false
+    protected final override val isBufferEmpty: Boolean get() = lock.withLock { value === EMPTY }
```

Comment:
The lock on `isBufferEmpty` should not be taken here, but in overrides for `isEmpty` and for `enqueueReceiveInternal` instead (see how `ArrayChannel` does it). This way, `isEmpty` will be linearizable, too.